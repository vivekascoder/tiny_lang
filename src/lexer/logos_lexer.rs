use logos::Logos;

#[derive(Logos, Debug, PartialEq)]
#[logos(skip r"([ \t\r\n]*|//[^\n]*)")]
enum Token {
    #[regex("[a-zA-Z_][a-zA-Z_0-9]*")]
    Ident,
    #[regex(r#""(?:\\.|[^"])*""#)]
    String,
    #[regex("-?[0-9]+")]
    Number,
    #[regex("(true|false)")]
    Boolean,
    #[token(":")]
    Colon,
    #[token(";")]
    SemiColon,

    // Operators
    #[token("=")]
    Equal,
    #[token("==")]
    DoubleEqual,
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Multiply,
    #[token("/")]
    Divide,
    #[token("%")]
    Mod,
    #[token("!=")]
    NotEqual,
    #[token("<")]
    LessThan,
    #[token("<=")]
    LessThanEqual,
    #[token(">")]
    GreaterThan,
    #[token(">=")]
    GreaterThanEqual,
    #[token("!")]
    Bang,
    #[token("(")]
    LParen,
    #[token(")")]
    RParen,
    #[token("{")]
    LBrace, // {
    #[token("}")]
    RBrace, // }
    #[token(",")]
    Comma,
    #[token("'")]
    SQuote, // '
    // #[token("\"")]
    // DQuote, // "
    #[token("\\")]
    BSlash, // \
    #[token("<<")]
    LeftShift, // <<
    #[token(">>")]
    RightShift, // >>
    #[token("&")]
    Ampersand, // &
    #[token("^")]
    Carrot, // ^
    #[token("|")]
    Pipe, // |

    // Keywords
    #[token("let")]
    KeywordLet,
    #[token("usize")]
    KeywordUsize,
    #[token("isize")]
    KeywordIsize,
    #[token("bool")]
    KeywordBool,
    #[token("fun")]
    KeywordFun,
    #[token("return")]
    KeywordReturn,
    #[token("if")]
    KeywordIf,
    #[token("else")]
    KeywordElse,
    #[token("void")]
    KeywordVoid,
    #[token("char")]
    KeywordChar,
    #[token("while")]
    KeywordWhile,

    // Symbol
    #[token("=>")]
    SymbolReturn,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_logos_lexer() {
        let mut lex = Token::lexer("let =");

        loop {
            if let Some(v) = lex.next() {
                println!("> {:?}, {:?}", v, lex.span());
            } else {
                break;
            }
        }
    }

    #[test]
    fn test_function_decl_and_call_logos() {
        let code = r#"
        let a = 445;
        let b = 45;
        let bool_var = false;
        let string_val = "this is a \"string";
        // let char_val = 'a';


        fun sum(a: usize, b: usize) => usize {
            let c = a + b;
            return c;
        }

        sum(a, b);
        "#;
        insta::assert_debug_snapshot!(Token::lexer(code).collect::<Vec<_>>());
    }
}
