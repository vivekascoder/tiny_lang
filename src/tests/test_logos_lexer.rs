use crate::lexer::logos_lexer::TokenKind;
use logos::Logos;

#[test]
fn test_logos_lexer() {
    let mut lex = TokenKind::lexer("let =");

    loop {
        if let Some(v) = lex.next() {
            println!("> {:?}, {:?}", v, lex.span());
        } else {
            break;
        }
    }
}

#[test]
fn test_function_decl_and_call_logos() {
    let code = r#"
    let a = 445;
    let b = 45;
    let bool_var = false;
    let string_val = "this is a \"string";
    let char_val = '\n';


    fun sum(a: usize, b: usize) => usize {
        let c = a + b;
        return c;
    }

    sum(a, b);
    "#;
    insta::assert_debug_snapshot!(TokenKind::lexer(code).collect::<Vec<_>>());
}
