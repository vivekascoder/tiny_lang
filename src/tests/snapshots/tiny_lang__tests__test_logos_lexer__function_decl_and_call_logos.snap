---
source: src/tests/test_logos_lexer.rs
expression: "TokenKind::lexer(code).collect::<Vec<_>>()"
---
[
    Ok(
        KeywordLet,
    ),
    Ok(
        Ident,
    ),
    Ok(
        Equal,
    ),
    Ok(
        Number(
            445,
        ),
    ),
    Ok(
        SemiColon,
    ),
    Ok(
        KeywordLet,
    ),
    Ok(
        Ident,
    ),
    Ok(
        Equal,
    ),
    Ok(
        Number(
            45,
        ),
    ),
    Ok(
        SemiColon,
    ),
    Ok(
        KeywordLet,
    ),
    Ok(
        Ident,
    ),
    Ok(
        Equal,
    ),
    Ok(
        Bool(
            false,
        ),
    ),
    Ok(
        SemiColon,
    ),
    Ok(
        KeywordLet,
    ),
    Ok(
        Ident,
    ),
    Ok(
        Equal,
    ),
    Ok(
        Str(
            "this is a \\\"string",
        ),
    ),
    Ok(
        SemiColon,
    ),
    Ok(
        KeywordLet,
    ),
    Ok(
        Ident,
    ),
    Ok(
        Equal,
    ),
    Ok(
        Char(
            '\n',
        ),
    ),
    Ok(
        SemiColon,
    ),
    Ok(
        KeywordFun,
    ),
    Ok(
        Ident,
    ),
    Ok(
        LParen,
    ),
    Ok(
        Ident,
    ),
    Ok(
        Colon,
    ),
    Ok(
        KeywordUsize,
    ),
    Ok(
        Comma,
    ),
    Ok(
        Ident,
    ),
    Ok(
        Colon,
    ),
    Ok(
        KeywordUsize,
    ),
    Ok(
        RParen,
    ),
    Ok(
        SymbolReturn,
    ),
    Ok(
        KeywordUsize,
    ),
    Ok(
        LBrace,
    ),
    Ok(
        KeywordLet,
    ),
    Ok(
        Ident,
    ),
    Ok(
        Equal,
    ),
    Ok(
        Ident,
    ),
    Ok(
        Plus,
    ),
    Ok(
        Ident,
    ),
    Ok(
        SemiColon,
    ),
    Ok(
        KeywordReturn,
    ),
    Ok(
        Ident,
    ),
    Ok(
        SemiColon,
    ),
    Ok(
        RBrace,
    ),
    Ok(
        Ident,
    ),
    Ok(
        LParen,
    ),
    Ok(
        Ident,
    ),
    Ok(
        Comma,
    ),
    Ok(
        Ident,
    ),
    Ok(
        RParen,
    ),
    Ok(
        SemiColon,
    ),
]
